### ISSUE TITLE
some questions

### ISSUE THREAD

**Issue by danielhrisca** [frejanordsiek/hdf5storage#73]
Hello Freja,

thank you for the nice work done with this project!

I've looked at some optimizations that could improve the time needed to write **mat** files. I had to cut some corners and the dirty result is found here https://github.com/danielhrisca/hdf5storage/tree/0.1.x (on my test file of ~210MB and 36000 datasets the _savemat_ execution time went from 72s to 37s)

Please consider the following questions
1. why do you need to make a deepcopy here https://github.com/frejanordsiek/hdf5storage/blob/0.1.x/hdf5storage/__init__.py#L1051? For 36400 objects this took 9s on my PC, and without deepcopy it takes 32ms

2. for write only mode some check can be skipped https://github.com/frejanordsiek/hdf5storage/compare/0.1.x...danielhrisca:0.1.x#diff-1c66dddc4eb3e5bdafdd75ba7095b7e9L1605 since the paths will not exist for sure in the file. Do you think you could add a new argument/attribute that could flag this mode and allow for a proper handling of such shortcuts?

3. there are a lot off h5py _\_\_getitem\_\__ calls caused by the frequent use of **grp[name]** instead of keeping a reference and reusing it. I've measured 25s spent on this, and using the reference practically took the time to 0s.

4. a refactoring using defaultdict (https://github.com/frejanordsiek/hdf5storage/compare/0.1.x...danielhrisca:0.1.x#diff-f20fd8a3cb7c00a2e178d8ac222f1755L1187) for the _writes_ function also improved the execution time, again because it avoided frequent _\_\_getitem\_\__ calls

**Comment by frejanordsiek** [comment]
Thank you for looking at this.

As for item 1, I see no reason it couldn't be changed (I did in fact change it the master branch https://github.com/frejanordsiek/hdf5storage/blob/b50295338625222869e70e8d82709c3750baddaa/hdf5storage/__init__.py#L1415). The API doesn't promise that it is a deep copy or anything in the 0.1.x series, so no one should be depending on it. The deep copy isn't needed. I just thought it might be when I originally wrote that code as a Python beginner back in 2013.

As for item 2, there is a plan to have such a flag to avoid those checks as well as https://github.com/frejanordsiek/hdf5storage/blob/a63682a7cd8e74d68ed5e3316336b209f6f8e30d/hdf5storage/utilities.py#L78 (with that function just using incrementing names instead of random ones) and https://github.com/frejanordsiek/hdf5storage/blob/a63682a7cd8e74d68ed5e3316336b209f6f8e30d/hdf5storage/utilities.py#L749 ; both of which consume significant time. The plan for the master branch is to make a class for an hdf5storage file that will have such a flag and the existing functions just make it and call the respective functions. Doing the same thing for the 0.1.x branch would not be possible without changing the API. There might be a way, however, to do it for the 0.1.x branch without changing the API. I am not sure how best to do it, but I will think about it.

As for item 3, they indeed are expensive and I had cut them out of the master branch. I had tried removing them from the 0.1.x branch but for some reason it didn't improve the benchmark I was using at all and in fact made it slightly worse (see Issue #42). But, the benchmark I was using was just one benchmark and it didn't hurt performance much. But it makes a huge difference in the one that you did. I am going to change it now. Going to do it a little bit differently than you did in your fork since API compatibility must be maintained for the 0.1.x branch and a little bit differently than I did for the master branch (your changes actually gave me a good idea for how to bypass one issue).

As for item 4, that is a good catch.


I am going to include the performance enhancements in items 1, 3, and 4 in the next bug fix release which will be on the sooner side since I just found a new bug I am getting close to resolving. Going to continue to think on item 2.
